Update Interaction and battle system

Ku6epnec opened the pull request:
Ветку перед заливом тестил в основном на дневной локации, в ночной может что-то не работать.

Ku6epnec committed:
Add Health/Damage Interfaces

Ku6epnec committed:
PlayerView update

Add Health, ProgressBar, Destroy and recieved damage

Ku6epnec committed:
Fix Attack Zombie Mod

Ku6epnec committed:
Add LookAt on PlayerAttack

Ku6epnec committed:
CageInteractive

Ku6epnec committed:
Change AreaCollider

Ku6epnec committed:
Add armor

Ku6epnec committed:
Fix bug when Player attack DeadZombie

Ku6epnec committed:
Delete Debug.Logs

Ku6epnec committed:
Add link on zombiePrefab

Ku6epnec committed:
Add time to attack

Ku6epnec committed:
Add AttackCollider on SunTime

Ku6epnec committed:
Add Interactive with Cage

Ku6epnec committed:
AxePrefab & delete debugs

Ku6epnec committed:
Fix prefabs

AlenaKarp commented on Assets/Art/Prototype/SkythianCat/Hand_Painted_Nature_Kit_LITE/DEMO/Demo_scene.unity:
Шо там сохранено на одной сцене и не сохранено на другой? Это бы вынести в Root префаб

Ku6epnec replied:
Там сохранено в префаб=\ просто когда тестил на сцене вначале закидывал на саму сцену, но в префаб тоже добавил, на сцене так и остался "+" вот он сохранил инфу, "+" удалил, оно само обновилось.

AlenaKarp commented on Assets/Scripts/AI/Zombie.cs:
```diff
@@ -28,15 +43,26 @@ public class Zombie : MonoBehaviour
         private ZombieState currentState;
 
+        private float timer;
+        private float Maxtimer = 1.0f;
```
- [x] Можно переименовать Maxtimer в maxTimer?

AlenaKarp commented on Assets/Scripts/AI/Zombie.cs:
```diff
@@ -28,15 +43,26 @@ public class Zombie : MonoBehaviour
 
         private void Awake()
         {
+            playerView = FindObjectOfType<ZombieFarm.Views.Player.PlayerView>();
```
- [x] Сделай на это дело ViewManager, который будет родителем над этими двумя ребятами, в этом менеджере в SerializedField на них ссылки, и публичные методы GetPlayerView() и GetInteractiveArea(), которые будут их возвращать соответственно (и вынести их в интерфейс IViewManager). У нас есть синглтон, который сейчас везде шарит менеджеры - Root. Туда надо будет прописать IViewManager. Таким образом тут не надо будет эти поля, а этих ребят чисто получать через Root.ViewsManager.GetPlayerView() и Root.ViewsManager.GetInteractiveArea(). Если над с этим помощь - пиши

Ku6epnec replied:
Сделалъ

AlenaKarp commented on Assets/Scripts/AI/Zombie.cs:
```diff
@@ -74,19 +107,30 @@ private void Chase()
+            playerView.RecievedDamage(Damage);
+        }
+
+        public void RecievedDamage(float damage)
```
- [x] Received, очепятка

AlenaKarp commented on Assets/Scripts/AI/Zombie.cs:
```diff
@@ -74,19 +107,30 @@ private void Chase()
+
+        public void RecievedDamage(float damage)
+        {
+            if (recievedDamageTimer <= 0)
```
- [x] И этой переменной тож опечатка

AlenaKarp commented on Assets/Scripts/Animal/Cage.cs:
```diff
@@ -4,22 +4,37 @@
+
+    private float _damage = 1;
+
+    private float Maxtimer = 2.0f;
```
maxTimer и received

AlenaKarp commented on Assets/Scripts/Managers/AttackTrigger.cs:
```diff
@@ -0,0 +1,43 @@
+        public Cage cage;
+        private void Awake()
+        {
+            playerView = FindObjectOfType<PlayerView>();
```
Root.ViewsManager.GetPlayerView()

AlenaKarp commented on Assets/Scripts/Managers/AttackTrigger.cs:
```diff
@@ -0,0 +1,43 @@
+        internal event Action ApplyDamage = () => { };
+        private PlayerView playerView;
+        public ZombieFarm.AI.Zombie Enemy;
+        public Cage cage;
```
Как будто бы все публичное может спокойно быть приватным. Если оно нужно где-то вне - надо сделать интерфейс и передавать через него через геттер. Плюс можно сделать интерфейс для enemy и cage - IInteractiveObject - и поля InteractiveObject, Enemy и Cage можно будет сократить до одного IInteractiveObject interactiveObject. Зомби ли это или клетка у тебя уже проверяется и передается в триггер энтере

AlenaKarp commented on Assets/Scripts/Managers/AttackTrigger.cs:
```diff
@@ -0,0 +1,43 @@
+            if (other.tag == "Zombie")
+            {
+                InteractiveObject = other.gameObject;
+                if (InteractiveObject.TryGetComponent<ZombieFarm.AI.Zombie>(out ZombieFarm.AI.Zombie zombie))
```
if (other.tag == "Zombie") и if (InteractiveObject.TryGetComponent<ZombieFarm.AI.Zombie>(out ZombieFarm.AI.Zombie zombie)) - масло маслянное, надо оставить что то одно, лучше второе

AlenaKarp commented on Assets/Scripts/Managers/Interfaces/IHaveInventory.cs:
Это кажется лишнее?

AlenaKarp commented on Assets/Scripts/Managers/InteractiveArea.cs:
```diff
@@ -0,0 +1,71 @@
+        internal event Action DeInteractive = () => { };
+        public LookAt lookAt;
+
+        private void OnTriggerEnter(Collider other)
```
У тебя два метода реализуют одно и то же при OnTriggerEnter'e. Это плохо, DRY-принцип, все дела. Я еще не дошла до того куда шарится InteractiveObject, но это есть не очень хорошо. Можно, например передавать их все в виде интерфейса как писала выше, и на стороне приема чекать что это за интерактивный объект и делать с ним дела.

Ku6epnec replied:
Решил вопрос при добавлении класса ReceivedDamageObject.

AlenaKarp replied:
А InteractiveObject всё еще public и шарится на сторону. (и lookAt) тоже. Если они нужны как public - надо хотя бы геттеры сделать, но в идеале чтобы они не выставлялись извне.

AlenaKarp commented on Assets/Scripts/UI/ProgressBar.cs:
```diff
@@ -13,26 +13,24 @@ public class ProgressBar : MonoBehaviour
-    private float middleLaneWidth;
+    float middleLaneWidth;
     private IEnumerator progress;
+    private float MaxHealth;
```
А если это будет прогресс бар не для здоровья? ;3
Мб назвать как-то поуниверсальнее, вроде maxValue? И стиль переменной тоже..

AlenaKarp commented on Assets/Scripts/UI/ProgressBar.cs:
```diff
@@ -13,26 +13,24 @@ public class ProgressBar : MonoBehaviour
     }
 
-    public void StartProgress()
+    public void StartProgress(float damage)
```
damage тоже лучше переназвать во что-то поуниверсальнее

AlenaKarp commented on Assets/Scripts/Views/Player/PlayerView.cs:
```diff
@@ -13,23 +13,72 @@ public class PlayerView : MonoBehaviour
+
+
+        public ZombieFarm.AI.Zombie enemy;
+        public Cage cage;
```
ъуъ писала про них выше

Ku6epnec replied:
Точно также засунул в ReceivedDamageObject.

AlenaKarp commented on Assets/Scripts/Views/Player/PlayerView.cs:
```diff
@@ -13,23 +13,72 @@ public class PlayerView : MonoBehaviour
+        public float Damage => _damage;
+
+        [Header("HealthStats")]
+        public float _health = 20;
```
Эти в SerializedField, у тебя они выше публично шарятся в интерфейсе, не надо дублировать

AlenaKarp commented on Assets/Scripts/Views/Player/PlayerView.cs:
```diff
@@ -13,23 +13,72 @@ public class PlayerView : MonoBehaviour
         }
 
-        private void OnAttack()
+        public void OnIdle()
```
Тут я понимаю оно зачем, править нинада, просто скажу что такой подход не очень хорош. Стейтами плеера должен руководить ток сам плеер, извне их менять кем-угодно плохая практика. (Да, я знаю что так уже сделано, просто если есть возможность так не делать - лучше не делать)

AlenaKarp commented on Assets/Scripts/Views/Player/PlayerView.cs:
```diff
@@ -13,23 +13,72 @@ public class PlayerView : MonoBehaviour
+            }
+            else if (cage != null)
+            {
+                cage.RecievedDamage(Damage);
```
А зачем нам промежуточный шаг в виде PlayerView? Почему бы не подписать на это событие сразу клетку и enemy? Мы ж их получаем в TriggerEnter в AttackTrigger'e и InteractiveArea, можно сразу и подписать

AlenaKarp commented on Assets/Scripts/Views/Player/PlayerView.cs:
```diff
@@ -46,12 +95,28 @@ public void DestroyObject(bool inProcess)
             }
         }
 
+        public void RecievedDamage(float damage)
```
received

AlenaKarp commented on Assets/Scripts/Views/Player/PlayerView.cs:
```diff
@@ -46,12 +95,28 @@ public void DestroyObject(bool inProcess)
             {
                 RefreshCurrentState(PlayerState.Idle);
-            }
+            }     
```
лишнее

Ku6epnec committed:
Fix any microelements

Ku6epnec committed:
Add polymorphism

Добавил класс ReceivedDamageObject, который теперь заменяет Монобехи на объектах, которые теоретически могут получать урон от игрока и в него засунул общий виртуальный метод Interaction. Теперь код в некоторых местах сильно упростился и стал удобнее, т.к. при различных тригерах персонаж просто проверяет есть ли у объекта перед ним данный класс, отказался от тэгов и убрал многое другое.

Ku6epnec committed:
Add ViewManager

Добавил ViewManager и GetPlayerView() с GetInteraciveArea()

Ku6epnec committed:
Fix nullification InteractiveObjetc

AlenaKarp commented on Assets/Scripts/Animal/LookAt.cs:
```diff
@@ -6,8 +6,20 @@ public class LookAt : MonoBehaviour
 {
     [SerializeField] GameObject _object;
```
А он нам еще нужен как [SerializeField], раз инитится через скрипт. Мб просто заприватить?

Ku6epnec replied:
Неее, у нас эта штука может и в дальнейшем использовать, когда нужно, чтобы кто-то или что-то пырилось на определённый объект и не всегда будет удобно через скрипт делать(особенно для тестов ГД), поэтому пускай будет. У нас же и коза тоже пырится на игрока, а не только игрок на зомби, уверен в дальнейшем ещё будут такие объекты попадаться.

AlenaKarp commented on Assets/Scripts/Managers/AttackTrigger.cs:
```diff
@@ -0,0 +1,35 @@
+{
+    public class AttackTrigger : MonoBehaviour
+    {
+        private GameObject InteractiveObject;
```
Поле теперь приватное - с маленькой буквы

AlenaKarp commented on Assets/Scripts/Managers/AttackTrigger.cs:
```diff
@@ -0,0 +1,35 @@
+        private GameObject InteractiveObject;
+        private PlayerView playerView;
+
+        private ReceivedDamageObject ReceivedDamageObject;
```
Поле теперь приватное - с маленькой буквы

Ku6epnec committed:
miniChanges

Исправил названия некоторых переменных

Ku6epnec committed:
Hide link InteractiveArea => Zombie

Теперь зомби не трогает данные InteractiveArea, а InteractiveArea при необходимости самостоятельно проверяет удалён ли интерактивный объект.

Ku6epnec committed:
Check CodeConvince

Отредактировал по КодКонвенции некоторые скрипты

AlenaKarp commented on Assets/Scripts/Managers/InteractiveArea.cs:
```diff
@@ -5,33 +5,36 @@ namespace ZombieFarm.Views.Player
+        private ReceivedDamageObject receivedDamageObject;
+        private LookAt lookAt;
+
+        internal event Action Interactive = () => { };
```
Туть тоже OnInteractive и OnDeinteractive

AlenaKarp commented on Assets/Scripts/Managers/InteractiveArea.cs:
```diff
@@ -38,24 +40,32 @@ private void OnTriggerExit(Collider other)
             }
         }
+
+        private void Cleaner()
```
Названия методов - глаголы. Clean()

Ku6epnec committed:
Rename Any Objects

Ku6epnec committed:
Delete any tags
